File: app/src/main/java/com/filemanager/fileexplorer/myfile/cleanmaster/free/Ultil/Image_Ultil.java
package com.filemanager.fileexplorer.myfile.cleanmaster.free.Ultil;

import android.annotation.SuppressLint;
import android.content.Context;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.provider.MediaStore;
import android.util.Log;
import android.widget.ImageView;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.IntentSenderRequest;

import com.filemanager.fileexplorer.myfile.cleanmaster.free.DTO.File_DTO;
import com.filemanager.fileexplorer.myfile.cleanmaster.free.MainActivity;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.text.SimpleDateFormat;
import java.util.ArrayList;


public class Image_Ultil {
    private Context context;
    private static ArrayList<File_DTO> file_dtos = new ArrayList<>();
    public static ArrayList<File_DTO> Album_iimage = new ArrayList<>();
    public static ArrayList<File_DTO> image_list = new ArrayList<>();
    public static long totolsize = 0;

    public Image_Ultil(Context context) {
        this.context = context;
    }
    @SuppressLint("Range")
    private ArrayList<File_DTO> getAllShownImagesPath() {
        ArrayList<File_DTO> arrayList = new ArrayList<>();
        Cursor query = this.context.getContentResolver().query(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, new String[]{"bucket_display_name", "_size", "_data", "_id", "date_added", "title", "_display_name", "date_modified"}, null, null, "date_added ASC");
        while (query.moveToNext()) {
            try {
                long j = query.getLong(query.getColumnIndexOrThrow("_size"));
                @SuppressLint("Range") String format = new SimpleDateFormat("dd/MM/yyyy | HH:mm:ss").format(Long.valueOf(query.getLong(query.getColumnIndex("date_modified")) * 1000));
                File_DTO file_DTO = new File_DTO();
                String string = query.getString(query.getColumnIndexOrThrow("_data"));
                String str = "" + query.getString(query.getColumnIndexOrThrow("_display_name"));
                file_DTO.setPath(string);
                if (str.equals("null")) {
                    file_DTO.setName(new File(string).getName());
                } else {
                    file_DTO.setName(str);
                }
                file_DTO.setSize(new Ultil(this.context).bytesToHuman(j));
                String str2 = "" + query.getString(query.getColumnIndexOrThrow("bucket_display_name"));
                if (str2.equals("null")) {
                    str2 = "Unknow";
                }
                file_DTO.setAbumname(str2);
                file_DTO.setDate(format);
                file_DTO.setId(query.getString(query.getColumnIndexOrThrow("_id")));
                totolsize += query.getLong(query.getColumnIndexOrThrow("_size"));
                arrayList.add(file_DTO);
            } catch (Exception unused) {
                Log.d("AAAA", "getAllShownImagesPath: ");
            }
        }
        return arrayList;
    }


    private ArrayList<File_DTO> getImageButket(Context context) {
        ArrayList<File_DTO> fileDTOList = new ArrayList<>();
        Uri uri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI;
        String[] projection = {"bucket_display_name", "_data", "date_modified"};

        try (Cursor cursor = context.getContentResolver().query(uri, projection, null, null, "date_added ASC")) {
            if (cursor != null) {
                while (cursor.moveToNext()) {
                    String albumName = cursor.getString((int)cursor.getColumnIndex(projection[0]));
                    String filePath = cursor.getString((int)cursor.getColumnIndex(projection[1]));
                    long dateModified = cursor.getLong((int)cursor.getColumnIndex(projection[2])) * 1000L;

                    String fileName = new File(filePath).getName();
                    String folderPath = filePath.replace("/" + fileName, "");

                    if (new File(filePath).exists()) {
                        File_DTO fileDTO = createFileDTO(fileDTOList, albumName, filePath, folderPath, dateModified);
                        fileDTOList.add(fileDTO);
                    }
                }
            }
        } catch (Exception e) {
            Log.e("TAG", "Exception is:  "+e);
            e.printStackTrace();
        }
        if (fileDTOList.isEmpty()) {
            Log.e("TAG", "fileDTOList is empty");
        }
        return fileDTOList;
    }

    private File_DTO createFileDTO(ArrayList<File_DTO> fileDTOList, String albumName, String filePath, String folderPath, long dateModified) {
        for (File_DTO fileDTO : fileDTOList) {
            if (fileDTO.getAbumname() != null && fileDTO.getAbumname().equals(albumName)) {
                int totalItems = fileDTO.getTotalitem();
                fileDTOList.remove(fileDTO);

                File_DTO updatedFileDTO = new File_DTO();
                updatedFileDTO.setAbumname(albumName);
                updatedFileDTO.setTotalitem(totalItems + 1);
                updatedFileDTO.setPath(filePath);
                updatedFileDTO.setRealpath(folderPath);
                updatedFileDTO.setDate(new SimpleDateFormat("dd/MM/yyyy | HH:mm:ss").format(dateModified));
                return updatedFileDTO;
            }
        }
        File_DTO newFileDTO = new File_DTO();
        newFileDTO.setAbumname(albumName);
        newFileDTO.setTotalitem(1);
        newFileDTO.setDate(String.valueOf(dateModified));
        newFileDTO.setPath(filePath);
        return newFileDTO;
    }
    public ArrayList<File_DTO> allImage() {
        image_list = getAllShownImagesPath();
        return image_list;
    }

    public ArrayList<File_DTO> updatedata() {
        image_list = new ArrayList<>();
        image_list = getAllShownImagesPath();
        return image_list;
    }

    public ArrayList<File_DTO> updatealbumimage() {
        Album_iimage = getImageButket(this.context);
        return Album_iimage;
    }

    public ArrayList<File_DTO> getallalbumImage() {
        if (Album_iimage.size() == 0 || Album_iimage.isEmpty()) {
            Album_iimage = getImageButket(this.context);
        }
        return Album_iimage;
    }

    public Bitmap getBitmap(String str, ImageView imageView) {
        File externalStorageDirectory = Environment.getExternalStorageDirectory();
        File file = new File(externalStorageDirectory + str, ".png");
        return Bitmap.createScaledBitmap(BitmapFactory.decodeFile(file.getAbsolutePath(), new BitmapFactory.Options()), imageView.getWidth(), imageView.getHeight(), true);
    }

    public ArrayList<File_DTO> delteabum(String str) {
        ArrayList<File_DTO> arrayList = new ArrayList<>();
        allImage();
        for (int i = 0; i < image_list.size(); i++) {
            if (image_list.get(i).getAbumname().equals(str)) {
                arrayList.add(image_list.get(i));
            }
        }
        return arrayList;
    }

    private void deletewithandroidversion(ArrayList<File_DTO> arrayList, ActivityResultLauncher<IntentSenderRequest> activityResultLauncher) {
        AndroidXI androidXI = new AndroidXI(this.context);
        Uri uri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI;
        if (Build.VERSION.SDK_INT < 30) {
            for (int i = 0; i < arrayList.size(); i++) {
                Uri withAppendedPath = Uri.withAppendedPath(uri, arrayList.get(i).getId());
                copyFile(arrayList.get(i).getPath());
                this.context.getContentResolver().delete(withAppendedPath, "_id=?", new String[]{arrayList.get(i).getId()});
            }
            return;
        }
        for (int i2 = 0; i2 < arrayList.size(); i2++) {
            Uri withAppendedPath2 = Uri.withAppendedPath(uri, arrayList.get(i2).getId());
            copyFile(arrayList.get(i2).getPath());
            androidXI.delete(activityResultLauncher, withAppendedPath2);
        }
    }

    private void copyFile(String str) {
        String replaceAll = str.replaceAll("/", "%");
        File file = new File(MainActivity.getStore(this.context) + "/Bin", replaceAll);
        try {
            InputStream openInputStream = this.context.getContentResolver().openInputStream(Uri.fromFile(new File(str)));
            FileOutputStream fileOutputStream = new FileOutputStream(file);
            MainActivity.copyStream(openInputStream, fileOutputStream);
            fileOutputStream.close();
            openInputStream.close();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
            Log.e("MTAG", "Image_Ultil.class,copyFile():" + e);
        } catch (IOException e2) {
            e2.printStackTrace();
            Log.e("MTAG", "Image_Ultil.class,copyFile():" +e2);

        }
    }
}

File: app/src/main/java/com/demo/filemanager/Ultil/Image_Ultil.kt
package com.demo.filemanager.Ultil

import android.app.Activity
import android.content.Context
import android.net.Uri
import android.provider.MediaStore
import android.util.Log
import com.demo.filemanager.DTO.File_DTO
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine


class Image_Ultil(private val context: Activity) {
    var Album_iimage = java.util.ArrayList<File_DTO>()
    companion object{
        var image_list = java.util.ArrayList<File_DTO>()
        var totolsize: Long = 0
    }
    suspend fun allImage(): ArrayList<File_DTO> = suspendCoroutine { continuation ->
        CoroutineScope(Dispatchers.Main).launch {
            // Call the function
            val result = getAllShownImagesPath()

            // Return the result
            continuation.resume(result as ArrayList<File_DTO>)
        }
    }

//    fun allImage(): ArrayList<File_DTO> {
//        image_list = getAllShownImagesPath()
//        return image_list
//    }


    suspend fun getAllShownImagesPath(): List<File_DTO> = withContext(Dispatchers.IO) {
        val arrayList = ArrayList<File_DTO>()

        context.contentResolver.query(
            MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
            arrayOf(
                "bucket_display_name",
                "_size",
                "_data",
                "_id",
                "date_added",
                "title",
                "_display_name",
                "date_modified"
            ),
            null,
            null,
            "date_added ASC"
        )?.use { query ->
            while (query.moveToNext()) {
                try {
                    val size = query.getLong(query.getColumnIndexOrThrow("_size"))
                    val modifiedDate = SimpleDateFormat("dd/MM/yyyy | HH:mm:ss").format(
                        query.getLong(query.getColumnIndexOrThrow("date_modified")) * 1000
                    )

                    val file_DTO = File_DTO().apply {
                        path = query.getString(query.getColumnIndexOrThrow("_data"))

                        val displayName =
                            query.getString(query.getColumnIndexOrThrow("_display_name"))
                        name = if (displayName == "null") File(path).name else displayName

                        val bucketName =
                            query.getString(query.getColumnIndexOrThrow("bucket_display_name"))
                        abumname = if (bucketName == "null") "Unknown" else bucketName

                        date = modifiedDate
                        id = query.getString(query.getColumnIndexOrThrow("_id"))
                    }

                    Image_Ultil.totolsize += size
                    arrayList.add(file_DTO)
                } catch (unused: Exception) {
                    Log.d("AAAA", "getAllShownImagesPath: ")
                }
            }
        }

        return@withContext arrayList
    }
//    suspend fun getAllShownImagesPath(): ArrayList<File_DTO> = withContext(Dispatchers.IO) {
//        val arrayList = ArrayList<File_DTO>()
//        val query = context.contentResolver.query(
//            MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
//            arrayOf(
//                "bucket_display_name",
//                "_size",
//                "_data",
//                "_id",
//                "date_added",
//                "title",
//                "_display_name",
//                "date_modified"
//            ),
//            null,
//            null,
//            "date_added ASC"
//        )
//
//        while (query!!.moveToNext()) {
//            try {
//                var size = query.getLong(query.getColumnIndexOrThrow("_size"))
//                @SuppressLint("Range") val format =
//                    SimpleDateFormat("dd/MM/yyyy | HH:mm:ss").format(
//                        java.lang.Long.valueOf(
//                            query.getLong(
//                                query.getColumnIndex("date_modified")
//                            ) * 1000
//                        )
//                    )
//
//                val file_DTO = File_DTO().apply {
//                    path = query.getString(query.getColumnIndexOrThrow("_data"))
//
//                    val displayName = query.getString(query.getColumnIndexOrThrow("_display_name"))
//                    name = if (displayName == "null") {
//                        File(path).name
//                    } else {
//                        displayName
//                    }
//
//                    val size = Ultil(context).bytesToHuman(size)
//
//                    var bucketName =
//                        query.getString(query.getColumnIndexOrThrow("bucket_display_name"))
//                    if (bucketName == "null") {
//                        bucketName = "Unknown"
//                    }
//                    abumname = bucketName
//
//                    date = format
//                    id = query.getString(query.getColumnIndexOrThrow("_id"))
//                }
//
//                Image_Ultil.totolsize += size
//                arrayList.add(file_DTO)
//            } catch (unused: Exception) {
//                Log.d("AAAA", "getAllShownImagesPath: ")
//            }
//        }
//
//        return@withContext arrayList
//    }

//    private fun getAllShownImagesPath(): ArrayList<File_DTO> {
//        val arrayList = ArrayList<File_DTO>()
//        val query = context.contentResolver.query(
//            MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
//            arrayOf(
//                "bucket_display_name",
//                "_size",
//                "_data",
//                "_id",
//                "date_added",
//                "title",
//                "_display_name",
//                "date_modified"
//            ),
//            null,
//            null,
//            "date_added ASC"
//        )
//        while (query!!.moveToNext()) {
//            try {
//                val j = query.getLong(query.getColumnIndexOrThrow("_size"))
//                @SuppressLint("Range") val format =
//                    SimpleDateFormat("dd/MM/yyyy | HH:mm:ss").format(
//                        java.lang.Long.valueOf(
//                            query.getLong(
//                                query.getColumnIndex("date_modified")
//                            ) * 1000
//                        )
//                    )
//                val file_DTO = File_DTO()
//                val string = query.getString(query.getColumnIndexOrThrow("_data"))
//                val str = "" + query.getString(query.getColumnIndexOrThrow("_display_name"))
//                file_DTO.path = string
//                if (str == "null") {
//                    file_DTO.name = File(string).name
//                } else {
//                    file_DTO.name = str
//                }
//                file_DTO.size = Ultil(context).bytesToHuman(j)
//                var str2 = "" + query.getString(query.getColumnIndexOrThrow("bucket_display_name"))
//                if (str2 == "null") {
//                    str2 = "Unknow"
//                }
//                file_DTO.abumname = str2
//                file_DTO.date = format
//                file_DTO.id = query.getString(query.getColumnIndexOrThrow("_id"))
//                Image_Ultil.totolsize += query.getLong(query.getColumnIndexOrThrow("_size"))
//                arrayList.add(file_DTO)
//            } catch (unused: Exception) {
//                Log.d("AAAA", "getAllShownImagesPath: ")
//            }
//        }
//        return arrayList
//    }

    suspend fun getallalbumImage(): java.util.ArrayList<File_DTO> {
        if (Album_iimage.size == 0 || Album_iimage.isEmpty()) {
            Album_iimage = getImageBucket(context) as java.util.ArrayList<File_DTO>
        }
        return Album_iimage
    }


    suspend fun getImageBucket(context: Context): List<File_DTO> = withContext(Dispatchers.IO) {
        val fileDTOList = mutableListOf<File_DTO>()
        val uri: Uri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI
//        val projection = arrayOf("bucket_display_name", "_data", "date_modified")
        val projection = arrayOf(MediaStore.Images.Media.BUCKET_DISPLAY_NAME, MediaStore.Images.Media.DATA, MediaStore.Images.Media.DATE_MODIFIED)
//        try {
            context.contentResolver.query(uri, projection, null, null, "date_added ASC")?.use { cursor ->
                while (cursor.moveToNext()) {
                    val albumName:String? = cursor.getString(cursor.getColumnIndex(MediaStore.Images.Media.BUCKET_DISPLAY_NAME)as Int)
                    val filePath = cursor.getString(cursor.getColumnIndex(MediaStore.Images.Media.DATA)as Int)
                    val dateModified = cursor.getLong(cursor.getColumnIndex(MediaStore.Images.Media.DATE_MODIFIED)as Int)

//                    val albumName = cursor.getString(cursor.getColumnIndex(projection[0]) as Int)
//                    val filePath = cursor.getString(cursor.getColumnIndex(projection[1])as Int)
//                    val dateModified = cursor.getLong(cursor.getColumnIndex(projection[2])as Int) * 1000L



                    val fileName = filePath.substringAfterLast('/')
                    val folderPath = filePath.replace("/$fileName", "")

                    if (java.io.File(filePath).exists()) {
                        val fileDTO = createFileDTO(fileDTOList,albumName?:"", filePath, folderPath, dateModified)
                        fileDTOList.add(fileDTO)
                    }
                }
            }
//        } catch (e: Exception) {
//            Log.e("TAG", "Exception is: $e")
//            e.printStackTrace()
//        }

        if (fileDTOList.isEmpty()) {
            Log.e("TAG", "fileDTOList is empty")
        }

        fileDTOList
    }




//    private fun createFileDTO(
//        fileDTOList: MutableList<File_DTO>,
//        albumName: String,
//        filePath: String,
//        folderPath: String,
//        dateModified: Long
//    ): File_DTO {
//        fileDTOList.find { it.abumname == albumName }?.let {
//            fileDTOList.remove(it)
//            return it.copy(totalitem = it.totalitem + 1, path = filePath, realpath = folderPath,
//                date = SimpleDateFormat("dd/MM/yyyy | HH:mm:ss", Locale.getDefault())
//                    .format(Date(dateModified)))
//        }
//        return File_DTO().apply {
//            this.abumname = albumName
//            this.totalitem = 1
//            this.date = SimpleDateFormat("dd/MM/yyyy | HH:mm:ss", Locale.getDefault())
//                .format(Date(dateModified))
//            this.path = filePath
//        }
//    }

    private fun createFileDTO(
        fileDTOList: MutableList<File_DTO>,
        albumName: String,
        filePath: String,
        folderPath: String,
        dateModified: Long
    ): File_DTO {
        fileDTOList.find { it.abumname == albumName }?.apply {
            val totalItems = this.totalitem
            fileDTOList.remove(this)
            return File_DTO().apply {
                this.abumname = albumName
                this.totalitem = totalItems + 1
                this.path = filePath
                this.realpath = folderPath
                this.date = SimpleDateFormat("dd/MM/yyyy | HH:mm:ss", Locale.getDefault())
                    .format(Date(dateModified))

            }
        }
        return File_DTO().apply {
            this.abumname = albumName
            this.totalitem = 1
            this.date = SimpleDateFormat("dd/MM/yyyy | HH:mm:ss", Locale.getDefault())
                .format(Date(dateModified))
            this.path = filePath
        }
    }

}